cirq-core/cirq/protocols/measurement_key_protocol.py

```diff
@@ -140,13 +155,67 @@ def measurement_keys(val: Any, *, allow_decompose: bool = True) -> AbstractSet[s
     return set()
 
 
-def is_measurement(val: Any) -> bool:
-    """Determines whether or not the given value is a measurement.
+def _is_measurement_from_magic_method(val: Any) -> Optional[bool]:
+    """Uses `is_measurement` magic method to determine if this object is a measurement."""
+    getter = getattr(val, '_is_measurement_', None)
+    return NotImplemented if getter is None else getter()
+
 
-    Measurements are identified by the fact that `cirq.measurement_keys` returns
-    a non-empty result for them.
+def _is_any_measurement(
+    vals: List[Any], vals_to_decompose: List[Any], allow_decompose: bool
+) -> bool:
+    """Given a list of objects, returns True if any of them is a measurement.
+
+    If `allow_decompose` is True, decomposes the objects and runs the measurement checks on the
+    constituent decomposed operations. But a decompose operation is only called if all cheaper
+    checks are done. A BFS for searching measurements, where "depth" is each level of decompose.
+    """
+    while vals:
+        val = vals.pop(0)
+        result = _is_measurement_from_magic_method(val)
+        if result is not NotImplemented:
+            if result is True:
+                return True
+            if result is False:
+                # Do not try any other strategies if `val` was explicitly marked as
+                # "not measurement".
+                continue
+
+        keys = _measurement_keys_from_magic_methods(val)
+        if keys is not NotImplemented and bool(keys) is True:
+            return True
+
+        if allow_decompose:
+            vals_to_decompose.append(val)
+
+        # If vals has finished iterating over, keep decomposing from vals_to_decompose until vals
+        # is populated with something.
+        while not vals:
```

Comment:
Any reason to prefer this structure over simply extending `vals` in the `if allow_decompose` block above?

Reply:
The assumption here is that decomposing is costly. So we are deferring any decomposing to only happen after we have exhausted all direct operation lookups. The method's docstring notes this.

Reply:
SGTM.